### PR TITLE
Show how to use 'require()' internally, bug fixes.

### PR DESCRIPTION
**Pull request by aisthesis**

### CODE REVIEW

**Comment by mscappini** [comment]
This

```
require('@optbot/template');
```

... is _not_ the same as this ...

```
var template = require('@optbot/template');
template();
```

They only happen to log the same output in this case because testMsg() is called from within the function definition, _as well as_ exported.

The first approach will only run once as Node actually caches the module definition once it is loaded. Requiring it again will not cause the app to output to the console more than once.

The second approach, however, creates a new copy of the exported function. So if you call `template();` twice, it will write to console twice.

**Comment by aisthesis** [comment]
They definitely aren't the same. I noticed when I tested the first version that with exactly that code, it wrote to console twice. I think the first is what we want, but we can discuss further tomorrow.

Here's where I'm coming from: `eqpuller` and `quotepuller` both depend on `pytools`. But once `pytools` is set up, you don't need to call it again to start and stop the dependent services. So, I was planning on having a setup script in each, and that would have `require('@optbot/pytools');`, which, it would appear, will actually execute the setup. Then you can start and stop the services without doing that again. I thought the more verbose version was necessary, but it looks like it isn't.

As to clarification of what you're saying: If I do

```
require('@optbot/template');
require('@optbot/template');
```

I'll only see the test message once. But if I use the original syntax, I'll see the message twice because it gets called on the first use of `require` and then again when I call the function that `require` returns.

My intention here in any case is just to provide the simplest possible example so that you can test your syntax for `require` on the `template` service before doing it for real on a more elaborate service.

**Comment by mscappini** [comment]
I'm still of the mindset that there shouldn't be anything too heavy that comes with `requiring` a package. I think it should be another exported function that gets called. It may be worth mentioning that if you feel the same.

**Comment by aisthesis** [comment]
I think the functionality of `require` should normally include what needs to be done _every time_ you would want to call `require`. If `main` needs to include functionality that you want to do sometimes but not always, then that shouldn't happen on `require`. `express` does far more on `require` than the model here does. Do you feel like they are doing things that are 'too heavy' on `require`? I don't like extra code for no reason.

What should imo _not_ execute on `require` is optional functionality. Anything that needs to happen every time can and should go in `main`.

**Comment by aisthesis** [comment]
Are you wanting to see the more verbose syntax in `template`?

**Comment by mscappini** [comment]
Actually, you'll notice that Express doesn't really do anything heavy when requiring a package. Most of what you see is only function definitions and setting those functions on the exports property. The heavy lifting comes from instantiating an instance of Express by doing this:

```
var express = require('express'),
    app = express();
```

That calls the `createApplication()` function to get the ball rolling by calling `app.init()` which sets up the default configuration. The first line where it `requires('express')` only defines those functions used.

**Comment by mscappini** [comment]
I think the circumstances vary per package, but ideally, I am still of the mindset that requiring something should not be an expensive operation. I see it as sort of importing a namespace. In the case of `pytools`, I don't think the expected behavior is to install tools on require. I would think that's an actionable item. If you want to avoid having it happen more than once in an app, you could easily store a variable to track that in the package.

**Comment by aisthesis** [comment]
in the case of pytools that is exactly what will happen. and if you call it with the verbose syntax, it will do it twice.

**Comment by mscappini** [comment]
Right, I am thinking you could just store a variable to prevent that from happening twice. Or you could just implement the logic you have to detect if the package does not need to be run.

**Comment by aisthesis** [comment]
i'm opposed to having more code than we really need. You're suggesting lots of work to avoid a nice, compact syntax.

**Comment by mscappini** [comment]
I believe what I am proposing requires very little code to accomplish what I consider to be a best practice. If we're constantly comparing our code with what other well established projects have done as a way to conform to supposed best practices, why is my confirmed refute for Express' position not a strong enough reason to alter this approach?

**Comment by aisthesis** [comment]
I don't view that really as a refute. `express` clearly does a lot more on `require` than our `template` does, including running through a `for` loop for a couple dozen iterations. Also, googling 'node require best practices', I don't see anything like that. [This](http://www.mircozeiss.com/node-js-require-s-best-practices/) is explicitly devoted to that and says nothing, but is the most prominent in google search.

I'm honestly not terribly passionate either way on this one in the sense that I'm not advocating that you always _must_ do it with `require` must always do something substantial. But it sounds like your objection is that you want a general prohibition on that, and I'm opposed to a general prohibition of being done with the dependency after just calling `require` on it. I think there are cases where that is desirable.

In other words, the more terse syntax shouldn't mean that that's the only way we do it. I actually agree that the other version is more common, because you want to expose more functionality than just the one thing. I could possibly expand the example to illustrate both approaches.

**Comment by mscappini** [comment]
Well, `express` does more than this template, but I am mostly taking about `pytools`. I know it's probably not the right place, but it's related. Express still only does things related to defining functions. None of the heavy stuff comes until you create an instance.

I am not for a general prohibition, I've been trying to be very clear about being against heavy lifting. I consider things like accessing network resources to be a heavy operation a la `pip install`. Or accessing drive resources like virtual environments.

I am also not against both approaches. I just want to keep heavy stuff out of the `require`.

**Comment by aisthesis** [comment]
Is there any more specific downside? Suppose something like `pip install` fails. I don't see why it's really going to matter whether that happens after:

```
require('@optbot/pytools');
```

or after

```
var pytools = require('@optbot/pytools');
pytools.setup();
```

I'll throw in an additional function to show both ways.

**Comment by mscappini** [comment]
If `setup()` does something actionable, there are a number of things that can occur that follow the general flow of how an app progresses.

Let's say you wanted to give `setup` some options. I think this looks better:

```
var http = require('http');
var pytools = require('@optbot/pytools');
var optbotPackage2 = require('@optbot/package2');

pytools.setup({
    option1: 'foo',
    callback: function() {
        console.log('something important');
    }
});
```

Than this:

```
var http = require('http');
var pytools = require('@optbot/pytools')({
    option1: 'foo',
    callback: function() {
        console.log('something important');
    }
});
var optbotPackage2 = require('@optbot/package2');
```

Or let's say you want to throw an exception from within the actionable method.

This:

```
var http = require('http');
var pytools = require('@optbot/pytools');
var optbotPackage2 = require('@optbot/package2');

try {
    pytools.setup({
        option1: 'foo',
        callback: function() {
            console.log('something important');
        }
    });
} catch {
}
```

Looks better than this:

```
var http = require('http');
try {
    var pytools = require('@optbot/pytools')({
            option1: 'foo',
            callback: function() {
                console.log('something important');
            }
        });
} catch {
}
var optbotPackage2 = require('@optbot/package2');
```

Of course, one could argue that all of that could be mitigated by moving around the dependency order, or not throw exceptions within the package `require()`. But why limit it if it can be avoided entirely?

**Comment by aisthesis** [comment]
What your examples demonstrate to me is that you can in all of those cases do it both ways with no problems. I'm still not seeing why you want to outlaw one of those choices at this point. As you say, "why limit it...?"

**Comment by mscappini** [comment]
Sure, technically speaking, you _could_ do it either way. But, at least from what I believe, we are discussing how one way not only looks better for legibility, but also functions more consistently with other packages I have observed (and not just within the Node.js ecosystem).

**Comment by mscappini** [comment]
I think I've exhausted most of the points I have as to why I would like to avoid doing heavy lifting on `require()`. If we can't come to some sort of mutual understanding based on what has already been discussed, then we should probably just continue on. I don't want to hold up a merge because of some standard I feel applies here.

**Comment by mscappini** [comment]
I've thought of one more example to help illustrate my opinion on the matter, in case I am not expressing this correctly or well enough. Let's use the `path` module as an example, in various languages.

Node.js

```
var path = require('path');
// Now use path variable.
```

C#

```
using System.IO;
// Now use Path class
```

Java

```
import java.nio.file;
// Now use Path interface
```

Python

```
import os.path
// Now use os.path
```

In each of these scenarios, I would expect them all to behave relatively similarly. No accessing network resources, no setting up virtual environments, no downloading tools, etc. Nothing complicated. I would expect them to return me an object or something that I can then _use_ for those specific operations.

JavaScript and Node.js are loose enough and don't enforce us to follow that pattern, but my experience with other languages makes this a more familiar and understandable paradigm.

If you understand this completely, but disagree, then we can certainly move on without concerning ourselves with it. I just wanted to ensure that the point was clear in case we agree. Regarding this pattern, though, I will be following it with the code that I contribute.

**Comment by aisthesis** [comment]
I ran across [this](http://browsenpm.org/package.json#main) in trying to put together a version that we would both like. Maybe you're right that `require`, like `import` in other language shouldn't connect to outside resources. If you look at the reference in the `main` script of `express`, it actually doing a huge amount--essentially running through all of the code in the repo through internal `require` statements, but it's not connecting externally. More extreme cases are `numpy` and `pandas`, where the import statements actually take a fair while, but again it's because a lot is going on internally with code that is already present locally and not because you're connecting to the outside world.

I'll go with the policy of "no external connections when calling `require`". I don't see why that needs to involve more than 1 extra line of code, and it may make debugging easier if the operation fails.

**Comment by mscappini** [comment]
Correct. The `express` module does do a lot of work when you `require` it, but it's not really the type of work I am concerned about. As mentioned, it's mostly just going through to define functions and such. Even when requiring other modules, those modules don't do anything on require, other than to define their own functions to be used later when an instance of `express` is created and puts them to use.

I like the changes you have made. I concur that you really only need one line of code to differentiate between the requiring of the module and some actionable method you intend to invoke.

I've even seen some examples around where people require the package then use it:

```
var instanceOfPackage = require('package').instance;
```

I think the most recent example I saw of this is [yargs](https://www.npmjs.com/package/yargs#but-don-t-walk-the-plank-just-yet-there-be-more-you-can-do-short-options). It looks like this:

```
var argv = require('yargs').argv;
```

...but maybe unnecessary even for `pytools`.

**Comment by mscappini** [comment]
Merge it!
